File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiPager.cpp
#include "StdAfx.h"
#include "UiPager.h"

#define ThisMethod($x) &UiPager::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiPager::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiPager> c_obj;
	}

	void UiPager::DefineControl()
	{
		AutoAddMethod( SetContent );
		AutoAddMethod( GetContent, WrapObjectGeneric );

		AutoAddMethod( SetContentSize );
		AutoAddMethod( GetContentSize, WrapObjectGeneric );
		AutoAddMethod( GetRealContentSize, WrapObjectGeneric );

		AutoAddMethod( SetContentHorizontalAlign );
		AutoAddMethod( GetContentHorizontalAlign, WrapObjectGeneric );

		AutoAddMethod( SetContentVerticalAlign );
		AutoAddMethod( GetContentVerticalAlign, WrapObjectGeneric );

		AutoAddMethod( SetAdjustment );
		AutoAddMethod( GetAdjustment, WrapObjectGeneric );

		AutoAddMethod( SetAutoHideScroll );
		AutoAddMethod( GetAutoHideScroll, WrapObjectGeneric );

		AutoAddMethod( SetPointerScroll );
		AutoAddMethod( GetPointerScroll, WrapObjectGeneric );

		AutoAddMethod( SetScrollPosition );
		AutoAddMethod( GetScrollPosition, WrapObjectGeneric );
		AutoAddMethod( GetScrollSize, WrapObjectGeneric );
		AutoAddMethod( GetScrollMax, WrapObjectGeneric );
	}

	U1 UiPager::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		return true;
	}

	WrapData<S32_2> UiPager::GetScrollMax() const
	{
		S32_2 v{};
		if ( GetControlTyped().GetScrollH().GetEnable() )
			v.x = GetControlTyped().GetScrollH().GetMaximum();
		if ( GetControlTyped().GetScrollV().GetEnable() )
			v.y = GetControlTyped().GetScrollV().GetMaximum();
		return v;
	}

}

File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Control/UiPager.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { WindowLike } from "./UiWindow";
import { Vec2 } from "../../Math/Vector";
import { AlignType } from "../UiCommon";

export enum PagerAdjustment {
	None,
	FitWidth,
	FitHeight,
}

export interface IPager extends IControl {
	new (window: WindowLike): IPager;

	SetContent(pControl: IControl): IControl;
	GetContent(): IControl;

	SetContentSize(vSize: Vec2): Pager;
	GetContentSize(): Vec2;
	GetRealContentSize(): Vec2;

	SetContentHorizontalAlign(nType: AlignType): Pager;
	GetContentHorizontalAlign(): AlignType;

	SetContentVerticalAlign(nType: AlignType): Pager;
	GetContentVerticalAlign(): AlignType;

	SetAdjustment(nAdjust: PagerAdjustment): Pager;
	GetAdjustment(): PagerAdjustment;

	SetAutoHideScroll(b: boolean): Pager;
	GetAutoHideScroll(): boolean;

	SetPointerScroll(b: boolean): Pager;
	GetPointerScroll(): boolean;

	SetScrollPosition(vPos: Vec2, bScroll: boolean): Pager;
	GetScrollPosition(): Vec2;
	GetScrollSize(): Vec2;
	GetScrollMax(): Vec2;
}

export class Pager extends (AveLib.UiPager as IPager) {
	private m_Content: IControl;

	SetContent(pControl: IControl) {
		this.m_Content = pControl;
		return super.SetContent(pControl);
	}

	GetContentSize(): Vec2 {
		return Vec2.FromNative(super.GetContentSize());
	}

	GetRealContentSize(): Vec2 {
		return Vec2.FromNative(super.GetRealContentSize());
	}

	GetScrollPosition(): Vec2 {
		return Vec2.FromNative(super.GetScrollPosition());
	}

	GetScrollSize(): Vec2 {
		return Vec2.FromNative(super.GetScrollSize());
	}

	GetScrollMax(): Vec2 {
		return Vec2.FromNative(super.GetScrollMax());
	}
}
